### PR TITLE
Clusterizer and new pixel masking

### PR DESCRIPTION
**Pull request by laborleben**
This is a feature branch to add additional support for the new pixel_clusterizer 3.0.

Topics:
- adding `hit_analysis.generate_pixel_mask()` to generate pixel mask
- modifying `hit_analysis.cluster_hits()` to work with pixel masks
- modify unit tests and compare old and new pixel removing method
- modifying examples make use of the above changes,
  and change the behavior from removing noisy pixels to removing single pix cluster of noisy pixels

The new method of masking pixel is **faster** than the old method where a new hit table was generated.
The old method of removing noisy hits from the hit table is still available.

### CODE REVIEW

File: testbeam_analysis/hit_analysis.py [hunk 1]
```diff
@@ -63,22 +67,76 @@ def remove_noisy_pixels(input_hits_file, n_pixel, output_hits_file=None,
...
     abs_occ_threshold = threshold * std
     occupancy = np.ma.masked_where(difference > abs_occ_threshold, occupancy)
     logging.info('Removed %d hot pixels at threshold %.1f in %s', np.ma.count_masked(occupancy), threshold, input_hits_file)
-
     # Generate tuple col / row array of hot pixels, do not use getmask()
     noisy_pixels_mask = np.ma.getmaskarray(occupancy)
+
+    with tb.open_file(output_mask_file, 'w') as out_file_h5:
+        # Creating occupancy table without masking noisy pixels
+        occupancy_array_table = out_file_h5.create_carray(out_file_h5.root, name='HistOcc', title='Occupancy Histogram', atom=tb.Atom.from_dtype(occupancy.dtype), shape=occupancy.shape, filters=tb.Filters(complib='blosc', complevel=5, fletcher32=False))
```
Comment (DavidLP) [inline testbeam_analysis/hit_analysis.py]: Maybe it is more convenient to add the noisy / deactivated pixel information as a user attribute to the Hits node? Then we do not need to deal with an additional file and additional function parameters. Python objects can be easily pickled by pytables and are even visible in vitables.

File: testbeam_analysis/hit_analysis.py [hunk 2]
```diff
@@ -63,22 +67,76 @@ def remove_noisy_pixels(input_hits_file, n_pixel, output_hits_file=None,
...
+        noisy_pixels_table = out_file_h5.create_carray(out_file_h5.root, name=pixel_mask_name, title='Pixel Mask', atom=tb.Atom.from_dtype(noisy_pixels_mask.dtype), shape=noisy_pixels_mask.shape, filters=tb.Filters(complib='blosc', complevel=5, fletcher32=False))
+        noisy_pixels_table[:] = noisy_pixels_mask
+
+    if plot:
+        plot_noisy_pixels(input_mask_file=output_mask_file, pixel_size=pixel_size, dut_name=dut_name)
+
+    return output_mask_file
+
+
+def remove_noisy_pixels(input_hits_file, n_pixel, output_hits_file=None, pixel_size=None, threshold=10.0, filter_size=3, dut_name=None, plot=True, chunk_size=1000000):
```
Comment (DavidLP) [inline testbeam_analysis/hit_analysis.py]: I prefer to have one function for one task to keep it simple. I would just not remove noisy pixels anymore when we have a mask for them.

File: testbeam_analysis/hit_analysis.py [hunk 3]
```diff
@@ -146,7 +206,7 @@ def cluster_hits(input_hits_file, output_cluster_file=None, max_x_distance=3, ma
             for hits, _ in analysis_utils.data_aligned_at_events(input_file_h5.root.Hits, chunk_size=chunk_size, try_speedup=False):
                 if not np.all(np.diff(hits['event_number']) >= 0):
                     raise RuntimeError('The event number does not always increase. The hits cannot be used like this!')
-                __, cluster = clusterizer.cluster_hits(hits)  # Cluster hits
+                cluster_hits, cluster = clusterizer.cluster_hits(hits, noisy_pixels=noisy_pixels, disabled_pixels=disabled_pixels)  # Cluster hits
```
Comment (DavidLP) [inline testbeam_analysis/hit_analysis.py]: Why to use `cluster_hits`?

File: testbeam_analysis/examples/eutelescope.py [hunk 4]
```diff
@@ -75,28 +75,31 @@ def run_analysis():
...
         'input_hits_file': data_files[i],
         'n_pixel': n_pixels[i],
+        'pixel_mask_name': "NoisyPixelMask",
         'pixel_size': pixel_size[i],
         'threshold': 0.5,
         'dut_name': dut_names[i]} for i in range(0, len(data_files))]
     pool = Pool()
     for kwarg in kwargs:
-        pool.apply_async(hit_analysis.remove_noisy_pixels, kwds=kwarg)
+        hit_analysis.generate_pixel_mask(**kwarg)
```
Comment (DavidLP) [inline testbeam_analysis/examples/eutelescope.py]: I guess you deactivated mp for testing, no?

File: testbeam_analysis/examples/telescope_time_reference.py [hunk 5]
```diff
@@ -47,29 +47,36 @@ def run_analysis():
...
         'input_hits_file': data_files[i],
         'n_pixel': n_pixels[i],
+        'pixel_mask_name': "NoisyPixelMask",
         'pixel_size': pixel_size[i],
         'threshold': threshold[i],
         'dut_name': dut_names[i]} for i in range(0, len(data_files))]
     pool = Pool()
     for kwarg in kwargs:
-        hit_analysis.remove_noisy_pixels(**kwarg)
+        hit_analysis.generate_pixel_mask(**kwarg)
```
Comment (DavidLP) [inline testbeam_analysis/examples/telescope_time_reference.py]: I guess you deactivated mp for testing, no?

**Comment by DavidLP** [comment]
> The new method of masking pixel is faster than the old method where a new hit table was generated

No really. Calling the noise remove / mask pixel function is faster yes. But at the end it cannot be faster overall because every function using hits has to evaluate the noisy pixel mask now. Effectively this was once done before, against now where it has to be done many times.

File: testbeam_analysis/hit_analysis.py (hunk 3 above)
Comment (laborleben) [inline testbeam_analysis/hit_analysis.py]: One might dump cluster_hits to disk for later use in the analysis chain?

**Comment by laborleben** [comment]
Yes, in principle right, but so far we don't use the hit table anywhere else. We only use the cluster table for the following analysis steps.
As I commented before, why shouldn't we save the cluster hits table to disk? This would make sense and can be used together with the cluster table.

File: testbeam_analysis/hit_analysis.py [hunk 6]
```diff
@@ -209,25 +209,23 @@ def cluster_hits(input_hits_file, output_cluster_file=None, input_disabled_pixel
...
                     raise RuntimeError('The event number does not always increase. The hits cannot be used like this!')
-                cluster_hits, cluster = clusterizer.cluster_hits(hits, noisy_pixels=noisy_pixels, disabled_pixels=disabled_pixels)  # Cluster hits
-                if not np.all(np.diff(cluster['event_number']) >= 0):
+                cluster_hits, clusters = clusterizer.cluster_hits(hits, noisy_pixels=noisy_pixels, disabled_pixels=disabled_pixels)  # Cluster hits
+                if not np.all(np.diff(clusters['event_number']) >= 0):
                     raise RuntimeError('The event number does not always increase. The cluster cannot be used like this!')
-                cluster_table_out.append(cluster)
+                # create cluster hits table dynamically
+                if cluster_hits_table is None:
+                    cluster_hits_table = output_file_h5.create_table(output_file_h5.root, name='ClusterHits', description=cluster_hits.dtype, title='Cluster hits table', filters=tb.Filters(complib='blosc', complevel=5, fletcher32=False))
```
Comment (DavidLP) [inline testbeam_analysis/hit_analysis.py]: Why do you want to store this if we do not use it?

**Comment by DavidLP** [comment]
> As I commented before, why shouldn't we save the cluster hits table to disk? This would make sense and can be used together with the cluster table.

Where?

File: testbeam_analysis/hit_analysis.py (hunk 3 above)
Comment (DavidLP) [inline testbeam_analysis/hit_analysis.py]: For what kind of analysis e.g.?

**Comment by DavidLP** [comment]
> As I commented before, why shouldn't we save the cluster hits table to disk? This would make sense and can be used together with the cluster table.

To use both tables together is not trivial and requires special, algorithmically intelligent parallel looping with numba speed up to get hits of a cluster. I do not see a use case for this now.

**Comment by laborleben** [comment]
This is done! All features included as discussed earlier.

**Comment by laborleben** [comment]
The cluster hist table is **not** added by default. This is just optional.

**Pull request merged**